File: Py/stationary_nonstationary.py
!pip install statsmodels
!pip install arch
!pip install scikit-learn


import yfinance as yf
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from statsmodels.tsa.stattools import adfuller, kpss
from arch.unitroot import PhillipsPerron
from sklearn.linear_model import LinearRegression



# Step 1: Get AMZN price data-------------
start_date = "2017-01-01"
end_date = "2023-02-09"

amzn_prices = yf.download("AMZN", start=start_date, end=end_date)

# Step 2: Check if data is stationary or non-stationary with URT (unit root tests)-------------

# 2.1 Apply ADF test
adf_result = adfuller(amzn_prices['Adj Close'], autolag='AIC')
adf_pvalue = adf_result[1]

# Null Hypothesis: Non-stationary
# Alt hypothesis: Stationary
# If p-value > 0.05, accept null / reject alt
# If p-value < 0.05, reject null / accept alt

# 2.2 Apply PP test
pp_result = PhillipsPerron(amzn_prices['Adj Close'])
# Null Hypothesis: Non-stationary
# Alt hypothesis: Stationary
# If p-value > 0.05, accept null / reject alt
# If p-value < 0.05, reject null / accept alt

# 2.3 Apply KPSS test
kpss_result = kpss(amzn_prices['Adj Close'], regression='c')
kpss_pvalue = kpss_result[1]

# Null Hypothesis: Stationary
# Alt hypothesis: Non-stationary
# If p-value > 0.05, accept null / reject alt
# If p-value < 0.05, reject null / accept alt

# Step 3: Detrending Methods-------------

# 3.1 Differencing
amzn_diff = amzn_prices['Adj Close'].diff()[1:]

# 3.2 DFA 
amzn_dfa2 = fathon.DFA(amzn_prices['Adj Close'])

def detrended_fluctuation_analysis(timeseries, max_window_size):
    N = len(timeseries)
    F = np.zeros(N)
    for i in range(N):
        if i < 2 or i > max_window_size:
            continue
        x = timeseries[0:i+1]
        y = np.cumsum(x - np.mean(x))
        z = np.polyfit(range(i+1), y, 1)
        p = np.poly1d(z)
        F[i] = np.sqrt(np.mean((y - p(range(i+1)))**2))
    return F


amzn_adj = amzn_prices['Adj Close'].values
F = detrended_fluctuation_analysis(amzn_adj, 400)
amzn_dfa = pd.DataFrame({'box': range(1, len(F) + 1), 'DFA': F})
# Remove 0's
amzn_dfa = amzn_dfa[amzn_dfa['DFA'] != 0]


# Add linear regression line of best fit
reg = LinearRegression().fit(amzn_dfa[['box']], amzn_dfa[['DFA']])
reg_line = reg.predict(amzn_dfa[['box']])[:, 0]

# Calculate Slope
slope = reg.coef_[0][0]



# Step 4: Transformation Methods-------------

# 4.1 Square root transformation
amzn_sqrt = np.sqrt(amzn_prices['Adj Close'])

# 4.2 Log Transformation
amzn_log = np.log(amzn_prices['Adj Close'])

# Step 5: Plot the time series + analysis to visualize-----------
fig, ax = plt.subplots(3, 2, figsize=(12, 10))

# 5.1 Plot Original AMZN prices
ax[0, 0].plot(amzn_prices.index, amzn_prices['Adj Close'], label='AMZN adjusted close price')
ax[0, 0].set_title("AMZN 5yr price chart")
ax[0, 0].set_xlabel("Year")
ax[0, 0].set_ylabel("Price")

# 5.2 Plot the differenced time series
ax[0, 1].plot(amzn_diff.index, amzn_diff.values)
ax[0, 1].set_title("AMZN Differenced Detrended Results")
ax[0, 1].set_xlabel("Date")
ax[0, 1].set_ylabel("Difference Detrended")

# 5.3 Plot sqrt Transformation
ax[1, 0].plot(amzn_sqrt.index, amzn_sqrt.values)
ax[1, 0].set_title("AMZN SQRT Transformation Results")
ax[1, 0].set_xlabel("Date")
ax[1, 0].set_ylabel("Sqrt Transform")

# 5.4 Plot Log Transformation
ax[1, 1].plot(amzn_log.index, amzn_log.values)
ax[1, 1].set_title("AMZN LOG Transformation Analysis Results")
ax[1, 1].set_xlabel("Date")
ax[1, 1].set_ylabel("Log Transform")

# 5.5 Plot DFA slope
ax[2, 0].scatter(amzn_dfa['box'], amzn_dfa['DFA'], label='DFA values')
ax[2, 0].plot(amzn_dfa['box'], reg_line, 'r', label='Line of best fit')
ax[2, 0].legend()
ax[2, 0].set_xlabel('Box size (number of data points)')
ax[2, 0.set_ylabel('DFA values')
ax[2, 0].set_title('Detrended Fluctuation Analysis of Amazon Stock Price \n Slope: ' + str(slope))


plt.tight_layout()
plt.show()

File: Stationary_Non-Stationary.R
library(tseries) # analysis methods
library(tidyquant) # quant data
library(ggplot2) # to plot
library(cowplot) # to merge plots
library(dfa) # for DFA
library(mgcv) # for spline regression

# Step 1: Get AMZN price data-------------
start_date <- as.Date("2017-01-01")
end_date <- as.Date("2023-02-09")

amzn_prices <- tq_get("AMZN", from = start_date, to = end_date)



# Step 2: Check if data is stationary or non-stationary with URT (unit root tests)-------------

# 2.1 Apply ADF test
adf_test <- adf.test(amzn_prices$adjusted)

# 2.2 Apply PP test
pp_test <- pp.test(amzn_prices$adjusted)

# Null Hypothesis: Non-stationary
# Alt hypothesis: Stationary
# If p-value > 0.05, accept null / reject alt
# If p-value < 0.05, reject null / accept alt

# 2.3 Apply KPSS test
kpss_test <- kpss.test(amzn_prices$adjusted)

# Null Hypothesis: Stationary
# Alt hypothesis: Non-stationary
# If p-value > 0.05, accept null / reject alt
# If p-value < 0.05, reject null / accept alt


# Step 3: Detrending Methods-------------

# 3.1 Detrended fluctuation analysis (DFA)
amzn_dfa_detrended <- DFA(amzn_prices$adjusted) %>%
        as.data.frame()

# 3.2 Spline regression
amzn_spline <- gam(adjusted ~ date, data = amzn_prices)
amzn_spline_detrended <- residuals(amzn_spline) %>%
        as.data.frame() %>%
        cbind(as.Date(amzn_prices$date))
colnames(amzn_spline_detrended) <- c("adjusted_residuals", "date")

# 3.3 Differencing
amzn_diff <- diff(amzn_prices$adjusted) %>%
        as.data.frame()


amzn_diff <- cbind(amzn_prices$date[2:nrow(amzn_prices)], amzn_diff)
colnames(amzn_diff) <- c("date", "adjusted_diff")

# Step 4: Transformation Methods-------------

# 4.1 Square root transformation
amzn_sqrt <- sqrt(amzn_prices$adjusted) %>%
        as.data.frame() %>%
        cbind(amzn_prices$date)
colnames(amzn_sqrt) <- c("adjusted_sqrt", "date")

# 4.2 Log Transformation
amzn_log <- log(amzn_prices$adjusted) %>%
        as.data.frame() %>%
        cbind(amzn_prices$date)
colnames(amzn_log) <- c("adjusted_log", "date")

# Step 5: Plot the time series + analysis to visualize-----------

# 5.1 Plot Original AMZN prices
plot_amzn <- amzn_prices %>%
        ggplot(aes(x = date, y = adjusted)) +
        geom_line() +
        theme_classic() +
        labs(title = "AMZN 5yr price chart") +
        xlab("Year") +
        ylab("Price")

# 5.2 Plot DFA slope

plot_amzn_dfa <- as.data.frame(amzn_dfa_detrended) %>%
        ggplot(aes(x = boxe, y = DFA)) +
        geom_point() +
        geom_smooth(method = "lm", se = FALSE) +
        theme_classic() +
        labs(title = "AMZN Detrended Fluctuation Analysis Results (slope 0.48)") +
        xlab("Window Size") +
        ylab("Detrended Fluctuation")
slope_dfa <- coef(lm(boxe ~ DFA, data = amzn_dfa_detrended))[2]
# Please check out this DFA post to learn how to interpret the slope and the results

#5.3 Plot spline regression residuals
plot_amzn_spline <- amzn_spline_detrended %>%
        ggplot(aes(x = date, y = adjusted_residuals)) +
        geom_line() +
        theme_classic() +
        labs(title = "AMZN Spline Regression Analysis Results") +
        xlab("Date") +
        ylab("Spline Residuals")

# 5.4 Plot the differenced time series
plot_amzn_diff <- amzn_diff %>%
        ggplot(aes(x = date, y = adjusted_diff)) +
        geom_line() +
        theme_classic() +
        labs(title = "AMZN Differenced Detrended Results") +
        xlab("Date") +
        ylab("Detrended Fluctuation")

# 5.5 Plot sqrt Transformation
plot_amzn_sqrt <- amzn_sqrt %>%
        ggplot(aes(x = date, y = adjusted_sqrt)) +
        geom_line() +
        theme_classic() +
        labs(title = "AMZN SQRT Transformation Results") +
        xlab("Date") +
        ylab("Sqrt Transform")

# 5.6 Plot Log Transformation
plot_amzn_log <-amzn_log %>%
        ggplot(aes(x = date, y = adjusted_log)) +
        geom_line() +
        theme_classic() +
        labs(title = "AMZN LOG Transformation Analysis Results") +
        xlab("Date") +
        ylab("Log Transform")


amzn_overlay <- plot_grid(plot_amzn, plot_amzn_diff, plot_amzn_sqrt,plot_amzn_log, plot_amzn_dfa, ncol = 2)
